WalletWasabi/ExchangeRate/ExchangeRateProvider.cs

```diff
@@ -0,0 +1,45 @@
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+using ExchangeRateProviderInfo = (string Name, string ApiUrl, System.Func<string, decimal> Extractor);
+
+namespace WalletWasabi.ExchangeRate;
+
+public class ExchangeRateProvider
+{
+	private static ExchangeRateProviderInfo[] Providers = [
+		("Bitstamp", "https://www.bitstamp.net/api/v2/ticker/btcusd", XPath(".bid")),
+		("Blockchain", "https://blockchain.info/ticker", XPath(".USD.buy")),
+		("Coinbase", "https://api.coinbase.com/v2/exchange-rates?currency=BTC", XPath(".data.rates.USD")),
+		("CoinGecko", "https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&ids=bitcoin", XPath(".[0].current_price")),
+		("Coingate", "https://api.coingate.com/v2/rates/merchant/BTC/USD", XPath("$")),
+		("Gemini", "https://api.gemini.com/v1/pubticker/btcusd", XPath(".bid")),
+	];
+
+	public async Task<ExchangeRate> GetExchangeRateAsync(string providerName, CancellationToken cancellationToken)
+	{
+		var providerInfo = Providers.First(x => x.Name == providerName);
+		var url = new Uri(providerInfo.ApiUrl);
+
+#pragma warning disable RS0030 // Do not use banned APIs
+		using var httpClient = new HttpClient();
+		httpClient.BaseAddress = new Uri($"{url.Scheme}://{url.Host}");
+		httpClient.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue("WasabiWallet", Helpers.Constants.ClientVersion.ToString()));
+#pragma warning restore RS0030 // Do not use banned APIs
```

Comment:
This has to use Tor, it is just that it is not implemented. The reason is that I still didn't consider the identity to be used and also because I am considering using the WebProxy (that supports socks5 proxies) instead of using our own huge machinery.

What you mention about the header is true, some providers reject the request if there is no header and I used that one just for developing. I have to change it.

Reply:
ofc, look at what I did for Wasabito: https://github.com/lontivero/Wasabito/blob/1e18aa0e2332a14c69761edb43bb9d5eefe0f9a4/WalletWasabi.Fluent.Desktop/Program.cs